packages/circuit-ui/components/ModalContext/ModalContext.tsx

```diff
@@ -0,0 +1,206 @@
+
+const ModalContext = createContext<ModalContextValue>([[], () => {}]);
+
+export interface ModalProviderProps extends Omit<ReactModalProps, 'isOpen'> {
+  initialState?: ModalState[];
+}
+
+export const ModalProvider: FC<ModalProviderProps> = ({
+  children,
+  initialState,
+  portalClassName = 'ReactModalPortal',
+  htmlOpenClassName = 'ReactModal__Html--open',
+  ...defaultModalProps
+}) => {
+  const [modals, dispatch] = useStack<ModalState>(initialState);
+
+  const isOpen = modals.length > 0;
+
+  useEffect(() => {
+    const popModal = () => {
+      dispatch({ type: 'pop' });
+    };
+
+    if (isOpen) {
+      window.onpopstate = popModal;
+    } else {
+      window.onpopstate = null;
+    }
+
+    return () => {
+      window.onpopstate = null;
+    };
+  }, [dispatch, isOpen]);
+
+  // TODO: Not sure this even works or is still needed.
+  // useEffect(() => {
+  //   if (!IS_IOS) {
+  //     return undefined;
+  //   }
+
+  //   if (isOpen) {
+  //     noScroll.on();
+  //   } else {
+  //     noScroll.off();
+  //   }
+
+  //   return () => {
+  //     noScroll.off();
+  //   };
+  // }, [isOpen]);
+
+  return (
+    <ModalContext.Provider value={[modals, dispatch]}>
+      {children}
+
+      {modals.map(
+        ({ id, onClose, timeout, component: Component, ...modalProps }) => {
+          const handleClose = (event?: MouseEvent | KeyboardEvent) => {
+            if (onClose) {
+              onClose(event);
+            }
+            dispatch({ type: 'remove', id, timeout: Component.TIMEOUT });
+          };
+          return (
+            <Component
+              {...defaultModalProps}
+              {...modalProps}
+              key={id}
+              isOpen={!timeout}
+              onClose={handleClose}
+              portalClassName={portalClassName}
+              htmlOpenClassName={htmlOpenClassName}
+            />
+          );
+        },
+      )}
+
+      {isOpen && (
+        <Global
+          styles={css`
+            /* Remove scroll on the body when react-modal is open */
+            .${htmlOpenClassName} {
+              height: 100%;
+              overflow-y: hidden;
+              -webkit-overflow-scrolling: auto;
+            }
+            /* Enable keyboard navigation inside the modal, see https://github.com/reactjs/react-modal/issues/782 */
+            .${portalClassName} {
+              position: absolute;
+              height: 100%;
+              width: 100%;
+            }
+          `}
+        />
+      )}
+    </ModalContext.Provider>
+  );
+};
+
+export function createUseModal<T extends BaseModalProps>(
+  component: ModalComponent<T>,
+) {
+  return (): {
+    setModal: (props: Omit<T, 'isOpen'>) => void;
+    removeModal: () => void;
+  } => {
+    const id = useMemo(uniqueId, []);
+    const [modals, dispatch] = useContext(ModalContext);
+
+    const modal = useMemo(() => modals.find((m) => m.id === id), [id, modals]);
```

Comment:
Could we move this into removeModal to avoid running it in setModal?

Reply:
It doesn't make a difference since it would get recalculated anyway — either in the `useMemo` here or the `useCallback` for `removeModal`.